### PR TITLE
Improve errors if the ssh server is not accessible

### PR DESCRIPTION
**Pull request by pchico83**
Signed-off-by: Pablo Chico de Guzman <pchico83@gmail.com>

## Proposed changes
- Added a new phase to check container connectivity
- Improve error when ssh server is not accessible
- Retry for 5s the first connection to the ssh server

Screenshot 2020-11-11 at 22 25 57

### CODE REVIEW

File: cmd/up/up.go [hunk 1]
```diff
@@ -352,14 +352,18 @@ func (up *upContext) activate(isRetry, autoDeploy, build bool) error {
...
 	}
 
+	log.Success("Development container activated")
+
 	if err := up.forwards(ctx); err != nil {
-		return fmt.Errorf("couldn't forward traffic to your development container: %s", err.Error())
+		if err == errors.ErrSSHConnectError {
+			return up.checkOktetoStartError(ctx, "Failed to connect to your development container")
+		}
+		return fmt.Errorf("couldn't connect to your development container: %s", err.Error())
```
Comment (rberrelleza) [inline cmd/up/up.go:361]: is the output of `err.Error` useful? we can log just the text and throw the full error in a log.info

File: cmd/up/up.go [hunk 2]
```diff
@@ -906,6 +887,35 @@ func (up *upContext) runCommand(ctx context.Context) error {
...
+	userID := pods.GetDevPodUserID(ctx, up.Dev, up.Client)
+	if up.Dev.PersistentVolumeEnabled() {
+		if userID != -1 && userID != *up.Dev.SecurityContext.RunAsUser {
+			return errors.UserError{
+				E:    fmt.Errorf("User %d doesn't have write permissions for the %s directory", userID, up.Dev.MountPath),
+				Hint: fmt.Sprintf("Set 'securityContext.runAsUser: %d' in your okteto manifest\n    After that, run 'okteto down -v' to reset your development container and run 'okteto up' again", userID),
+			}
+		}
+	} else {
+		if pods.OktetoDevPodMustBeRecreated(ctx, up.Dev, up.Client) {
```
Comment (rberrelleza) [inline cmd/up/up.go:900]: it's odd that the function called `checkOktetoStartError` changes the system. can we split the destruction into a separate stage? (or return a speficic error, and then the mail loop does the recreation?

File: pkg/ssh/pool.go [hunk 3]
```diff
@@ -54,6 +56,33 @@ func startPool(ctx context.Context, serverAddr string, config *ssh.ClientConfig)
 	return p, nil
 }
 
+func retryNewClientConn(ctx context.Context, c net.Conn, addr string, conf *ssh.ClientConfig) (ssh.Conn, <-chan ssh.NewChannel, <-chan *ssh.Request, error) {
+	ticker := time.NewTicker(300 * time.Millisecond)
+	to := config.GetTimeout() / 6 // 5 seconds
```
Comment (rberrelleza) [inline pkg/ssh/pool.go:61]: could we just keep it for the full 30s ? it's the worst case scenario right?

File: pkg/ssh/pool.go [hunk 4]
```diff
@@ -54,6 +56,33 @@ func startPool(ctx context.Context, serverAddr string, config *ssh.ClientConfig)
...
+
+		if time.Now().After(timeout) {
+			return nil, nil, nil, err
+		}
+
+		select {
+		case <-ticker.C:
+			continue
+		case <-ctx.Done():
+			log.Infof("ssh.retryNewClientConn cancelled")
```
Comment (rberrelleza) [inline pkg/ssh/pool.go:80]: done returns an err (or you can get it from the ctx), I find it's useful to log it to know the reason of the cancellation.

**Review by rberrelleza — APPROVED** [review]

File: cmd/up/up.go (hunk 1 above)
Comment (pchico83) [reply cmd/up/up.go:361]: I find it useful. It is how it was before and people usually report this error without needing to send a doctor file

File: pkg/ssh/pool.go (hunk 3 above)
Comment (pchico83) [reply pkg/ssh/pool.go:61]: it is better to fail fast. the `ssh` server starts very fast, but I have found the first connection to fail a few times

File: pkg/ssh/pool.go (hunk 4 above)
Comment (pchico83) [reply pkg/ssh/pool.go:80]: ok

File: cmd/up/up.go (hunk 2 above)
Comment (pchico83) [reply cmd/up/up.go:900]: ok

**Pull request merged**